pkg/deployment/cert.go

```diff
@@ -138,20 +141,56 @@ func EnsureTLSCerts(ctx context.Context, helper *helper.Helper,
 		certsData[hostName+"-ca.crt"] = certSecret.Data["ca.crt"]
 	}
 
-	// create a secret to hold the certs for the service
-	serviceCertsSecret := &corev1.Secret{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      GetServiceCertsSecretName(instance, service.Name),
-			Namespace: instance.Namespace,
-		},
-		Data: certsData,
+	// Calculate number of secrets to create
+	ci := []map[string][]byte{}
+
+	keys := []string{}
+	for k := range certsData {
+		keys = append(keys, k)
 	}
-	_, result, err := secret.CreateOrPatchSecret(ctx, helper, instance, serviceCertsSecret)
-	if err != nil {
-		err = fmt.Errorf("error creating certs secret for %s - %w", service.Name, err)
-		return &ctrl.Result{}, err
-	} else if result != controllerutil.OperationResultNone {
-		return &ctrl.Result{RequeueAfter: time.Second * 5}, nil
+	sort.Strings(keys)
+
+	totalSize := secretMaxSize
+	var cur *map[string][]byte
+	// Going 3 by 3 to include ca, crt and key in the same secret.
+	for k := 0; k < len(keys)-1; k += 3 {
+		szCa := len(certsData[keys[k]]) + len(keys[k])
+		szCrt := len(certsData[keys[k+1]]) + len(keys[k+1])
+		szKey := len(certsData[keys[k+2]]) + len(keys[k+2])
+		sz := szCa + szCrt + szKey
+		if (totalSize + sz) > secretMaxSize {
+			i := len(ci)
+			ci = append(ci, make(map[string][]byte))
+			cur = &ci[i]
+			totalSize = 0
+		}
+		totalSize += sz
+		(*cur)[keys[k]] = certsData[keys[k]]
+		(*cur)[keys[k+1]] = certsData[keys[k+1]]
+		(*cur)[keys[k+2]] = certsData[keys[k+2]]
+	}
+
```

Comment:
I would prefer we move this code to a separate utility which possibly can be re-used and not clutter the `EnsureTLSCerts` function.

Reply:
As discussed out of band with @rabi, this is not a block for approving/merging this PR. The refactoring will be  done on a next patch.